narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,226 +3,381 @@
-        tokio::pin!(timer);
-        let mut attempt_garbage_collection;
-
-        info!("CertificateWaiter has started successfully.");
         loop {
-            // Initially set to not garbage collect
-            attempt_garbage_collection = false;
-
             tokio::select! {
-                // We only accept new elements if we have "room" for them
-                Some(certificate) = self.rx_synchronizer.recv(), if waiting.available_permits() > 0 => {
-                    if certificate.epoch() < self.committee.epoch() {
+                Some(certificate) = self.rx_certificate_waiter.recv() => {
+                    let header = &certificate.header;
+                    if header.epoch != self.committee.epoch() {
                         continue;
                     }
+                    // Unnecessary to validate the header and certificate further, since it has already been validated.
 
-                    // Ensure we process only once this certificate.
-                    let header_id = certificate.header.id;
-                    if self.pending.contains_key(&header_id) {
-                        continue;
+                    if let Some(r) = self.target.get(&header.author) {
+                        if header.round <= *r {
+                            // Ignore fetch request when we already need to sync to a later certificate.
+                            continue;
+                        }
                     }
 
-                    // Add the certificate to the waiter pool. The waiter will return it to us when
-                    // all its parents are in the store.
-                    let wait_for = certificate.header.parents.iter().cloned().collect();
-                    let (tx_cancel, rx_cancel) = oneshot::channel();
-                    // TODO: remove all this once drain_filter is stabilized.
-                    let once_cancel = {
-                        let inner = OnceCell::new();
-                        inner.set(tx_cancel).expect("OnceCell invariant violated");
-                        inner
+                    // The header should have been verified as part of the certificate.
+                    match self.certificate_store.last_round_number(&header.author) {
+                        Ok(Some(r)) => {
+                            if r >= header.round {
+                                // Ignore fetch request. Possibly the certificate was processed while the message is in the queue.
+                                continue;
+                            }
+                            // Otherwise, continue to update fetch target.
+                        },
+                        Ok(None) => {
+                            // The authority has no update since genesis. Continue to update fetch target.
+                        },
+                        Err(e) => {
+                            warn!("Failed to read latest round for {}: {}", header.author, e);
```

Comment:
Should this be an `error!`? i.e., is this never supposed to happen unless something is broken with our code? Or could it be caused by byzantine remote authority, etc?

Reply:
I thought the error would come from rocksdb, but actually it can only come from bincode serialization failures. Updated to use error and added a comment.